### PR TITLE
[TTPReq] Tracking

### LINKED ISSUE #14328: [TTPReq] Tracking
### Events

More context: pffQ75-16f-p2

### PR DESCRIPTION
**Pull request by staskus**
Closes: #14328
More context: pffQ75-16f-p2

## Description

Added tracking to the project. Main areas:
- Tap to Pay Awareness Modal
- Tap to Pay Education
- Card Reader Location Alerts
- Email Receipt View

I had to do a little refactoring around the views for easier life cycle tracking. The behavior is unchanged.

### Reusing existing events
Name | Event
-- | --
About Tap to Pay Selected | payments_hub_tap_to_pay_about_tapped
Set up Tap to Pay Selected | payments_hub_tap_to_pay_tapped
Set up Tap to Pay Flow Initiated | tap_to_pay_set_up_information_set_up_tapped
TTP Connected | card_reader_connection_successcard_reader_model:COTS_DEVICE
Email Receipt Option Tapped | receipt_email_tapped
Email Receipt View Dismissed | receipt_email_canceled
Email Receipt Succeeded | receipt_email_success
Email Receipt Failed | receipt_email_failed

### New events

I opened event registration PRs and I will merge them and validate events once this PR is reviewed.

Name | Event 
-- | --
TTP Education Steps Shown | tap_to_pay_education_shown
TTP Education Steps Completed | tap_to_pay_education_done
TTP Education Steps Skipped | tap_to_pay_education_skipped
TTP Terms of Services Accepted | tap_to_pay_terms_of_service_accepted
TTP Awareness Modal Shown | tap_to_pay_awareness_shown
Card Reader Payment Location Permission Pre-alert Shown | card_reader_location_permission_pre_alert_shown
Card Reader Payment Location Permission Required Alert Shown | card_reader_location_permission_required_shown

## Steps to reproduce

**Prerequisites**:
- Open a site with WooCommerce 9.5 and WooPayments 8.6 (both versions are officially released, so just update them to the latest version in plugins directory)
- Set `showingTapToPayAwarenessMoment` in `DashboardViewModel` to open awareness modal more easily
- Use a Tap to Pay eligible device or simulator
- Use a simulated card reader or remove your device for a fresh Tap to Pay setup https://businessconnect.apple.com/taptopay/removeall
- Go to 'Privacy & Security' -> 'Location Services' -> 'Woo' and set to 'Ask Next Time or When I Share'

All the events could be reproduced in a single flow (see video):

1. Open the app
2. Confirm TTP awareness modal is opened and `tap_to_pay_awareness_shown` tracked
3. Tap 'Learn More'
4 Confirm Tap to Pay education is opened and `payments_hub_tap_to_pay_about_tapped` tracked
5. Tap 'Skip', confirm `tap_to_pay_education_skipped` tracked
6. Tap `Set up Tap to Pay`, confirm `payments_hub_tap_to_pay_tapped` tracked
7. Tap `Set up Tap to Pay` again, confirm `tap_to_pay_set_up_information_set_up_tapped` tracked
8. When the Terms of Services is opened, accept it, confirm `tap_to_pay_terms_of_service_accepted` tracked
9. Education should open again, confirm `tap_to_pay_education_shown` tracked
10. Go through the flow and tap 'Done, confirm `tap_to_pay_education_done` tracked
11. Set up should finish, try a Tap to Pay transaction

Payment flow:
13. When location pre-alert appears, confirm `card_reader_location_permission_pre_alert_shown` tracked
14. Tap 'Continue' and deny the location request
15. When the location required alert appears, confirm `card_reader_location_permission_required_shown` tracked
16. Open device settings, set location to 'While Using the App'
17. Proceed through the payment flow
19. Confirm the payment alert has 'Email Receipt' option, tap it, confirm `receipt_email_tapped` tracked
20. Dismiss it with the cancel button, confirm `receipt_email_canceled` tracked, and select 'Email Receipt' again
21. Dismiss it with the down gesture, confirm `receipt_email_canceled` tracked, and select 'Email Receipt' again
22. Enter the email, turn off the internet connection, tap 'Send', confirm `receipt_email_failed` tracked
23. Turn on the internet, submit again, confirm `receipt_email_success` tracked 

## Testing information

This task touches the whole feature. I suggest focusing on the events since the feature itself and the corner cases will be tested thoroughly during the CfT.  In this task, I updated the unit tests required after refactoring. All the order tests continue to run successfully. 

## Video - full flow

https://github.com/user-attachments/assets/4782e021-1831-4bd9-8d4c-b9f959155051

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Experiments/Experiments/DefaultFeatureFlagService.swift [hunk 1]
```diff
@@ -92,11 +92,11 @@ public struct DefaultFeatureFlagService: FeatureFlagService {
         case .paymentsOnboardingInPointOfSale:
             return true
         case .sendReceiptAfterPayment:
-            return false
+            return buildConfig == .localDeveloper || buildConfig == .alpha
         case .sendReceiptsForPointOfSale:
             return false
         case .tapToPayEducation:
-            return false
+            return buildConfig == .localDeveloper || buildConfig == .alpha
```
Comment (jaclync) [inline Experiments/Experiments/DefaultFeatureFlagService.swift]: ❓ Do we want to just enable the feature flags for debug and installable builds? Or all users with `true`?

File: WooCommerce/Classes/Analytics/WooAnalyticsStat.swift [hunk 2]
```diff
@@ -347,15 +347,14 @@ enum WooAnalyticsStat: String {
     case cardReaderSelectTypeShown = "card_present_select_reader_type_shown"
     case cardReaderSelectTypeBuiltInTapped = "card_present_select_reader_type_built_in_tapped"
     case cardReaderSelectTypeBluetoothTapped = "card_present_select_reader_type_bluetooth_tapped"
-    case cardReaderDiscoveryTapped = "card_reader_discovery_tapped"
     case cardReaderDiscoveryFailed = "card_reader_discovery_failed"
-    case cardReaderDiscoveredReader = "card_reader_discovery_reader_discovered"
-    case cardReaderConnectionTapped = "card_reader_connection_tapped"
```
Comment (jaclync) [inline WooCommerce/Classes/Analytics/WooAnalyticsStat.swift:353]: Nice cleanup, I can't even see these deleted events in the Tracks directory. I see the equivalent events in Android with active numbers in the last two weeks, maybe there are some UX differences between platforms.

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/BuiltInCardReaderConnectionController.swift [hunk 3]
```diff
@@ -364,20 +364,22 @@ private extension BuiltInCardReaderConnectionController {
...
     ///
     func onRequestLocationPermission() {
+        // Refresh the view if the location permission state changes
+        locationService.observePermissionChanges { [weak self] permission in
+            guard let self else { return }
+            locationService.stopObservingPermissionChanges()
+            if case .requestLocationPermission = state {
+                onRequestLocationPermission()
+            }
+        }
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/CardPresentPayments/BuiltInCardReaderConnectionController.swift]: as a very unlikely edge case, could this be triggered for the .authorized case below? Maybe this could be in a private function and called by just the `denied` and `notDetermined` cases?

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/Tap to Pay Education/TapToPayEducationView.swift [hunk 4]
```diff
@@ -60,17 +60,12 @@ struct TapToPayEducationView: View {
...
         }
         .interactiveDismissDisabled(viewModel.isInteractiveDismissDisabled)
-        .sheet(isPresented: $viewModel.showingSetUpFlow,
-               onDismiss: viewModel.reloadHasPreviousTapToPayUsage,
-               content: {
-            TapToPaySettingsFlowPresentingView(
-                configuration: viewModel.configuration,
-                siteID: viewModel.siteID,
-                onboardingUseCase: viewModel.cardPresentPaymentsOnboardingUseCase)
-        })
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/Tap to Pay Education/TapToPayEducationView.swift:70]: just double-checking, is this deletion intentional?

File: WooCommerce/WooCommerceTests/Tools/LocationServiceTests.swift [hunk 5]
```diff
@@ -46,20 +38,10 @@ struct LocationServiceTests {
...
 
         // Then
-        #expect(status == nil)
         #expect(locationManager.requestWhenInUseAuthorizationCalled)
-
-        // When
-        locationManager.changeAuthorizationStatus(to: .authorizedWhenInUse)
-
-        // Then
-        #expect(status == .authorized)
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/Tools/LocationServiceTests.swift:62]: non-blocking: maybe the status assertions can be recovered in a different way?

File: WooCommerce/WooCommerceTests/ViewModels/Receipts/ReceiptEmailViewModelTests.swift [hunk 6]
```diff
@@ -9,6 +11,7 @@ struct ReceiptEmailViewModelTests {
     private let order: Order
     private let noticesPresenter: MockNoticePresenter
     private let sut: ReceiptEmailViewModel
+    private var cancellables = Set<AnyCancellable>()
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/ViewModels/Receipts/ReceiptEmailViewModelTests.swift]: nit: we started calling this set of subscriptions `subscriptions` instead of `cancellables` from maybe last year or so, it's not documented though.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift [hunk 7]
```diff
@@ -34,20 +46,57 @@ final class ReceiptEmailViewModel: ObservableObject {
...
+                }
             }
         }
 
         self.state = .loading
         stores.dispatch(action)
     }
+
+    func onAppear() {
+        analytics.track(event: .InPersonPayments.receiptEmailTapped(
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift]: nit: from this event name, it feels like it's triggered when a button is tapped. "receiptEmailShown" might sound more accurate here?

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift [hunk 8]
```diff
@@ -34,20 +46,57 @@ final class ReceiptEmailViewModel: ObservableObject {
...
+        )
+    }
+
+    func onDisappear() {
+        if state != .success {
+            analytics.track(event: .InPersonPayments.receiptEmailCanceled(
+                countryCode: countryCode,
+                cardReaderModel: cardReaderModel,
+                source: .api)
+            )
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift]: nit: can this be called in `onCancel`?

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift [hunk 9]
```diff
@@ -34,20 +46,57 @@ final class ReceiptEmailViewModel: ObservableObject {
...
+        }
+
+        self.noticePresenter.presentingViewController = nil
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift]: super nit: self probably not needed (same in `onCancel`)
```suggestion
        noticePresenter.presentingViewController = nil
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift [hunk 10]
```diff
@@ -1,24 +1,36 @@
...
     @Published var state: PrimaryLoadingButtonStyle.State = .idle
+    @Published private(set) var dismiss: Bool = false
 
-    private let order: Order
+    private var order: Order
     private let stores: StoresManager
+    private let analytics: Analytics
+    private let countryCode: CountryCode
+    private let cardReaderModel: String?
+    var onDismiss: (Order?) -> Void
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift]: nit: this could be let?
```suggestion
    let onDismiss: (Order?) -> Void
```

**Review by jaclync — COMMENTED** [review]
> 4. Tap 'Skip', confirm tap_to_pay_education_skipped tracked
> 5. Tap `Set up Tap to Pay`, confirm payments_hub_tap_to_pay_tapped tracked

After tapping ‘Skip’, there’s no CTA `Set up Tap to Pay`, just `Done`. In another store, I was able to see `Set Up Tap to Pay on iPhone` CTA and the `tap_to_pay_set_up_information_set_up_tapped` event.

> 6. Tap Set up Tap to Pay again, confirm tap_to_pay_set_up_information_set_up_tapped tracked
> 7. When the Terms of Services is opened, accept it, confirm tap_to_pay_terms_of_service_accepted tracked

I still couldn’t get the ToS to be shown again with a fresh install, https://businessconnect.apple.com/taptopay/removeall showed `There are no merchant IDs associated with this Apple Account` for my A8C Apple developer account. Do you see a different page?

Screenshot 2024-12-17 at 1 46 31 PM


---

Other than these two cases, the rest of the testing worked as expected. I tested on iPhone 13 Pro iOS 17.5.1. Lemme know if I missed something on the setup. For testing the awareness modal, I hard coded `TapToPayAwarenessMomentDeterminer.shouldPresent` for the modal to show up.

---

During testing, I noticed an issue:

- Menu > Payments > About Tap To Pay
- Tap `Next` until `Set Up Tap to Pay on iPhone` shows up
- Tap `Set Up Tap to Pay on iPhone`
- Tap `Set Up Tap to Pay on iPhone` again
- Tap `Continue`
- Tap `Not now`  --> after the view is dismissed, the `Setup complete` modal appears again which feels unexpected


https://github.com/user-attachments/assets/d938a478-95b4-4ec6-98a8-80e8bdb1b5da

---

A minor observation: `Set Up Tap to Pay on iPhone` and `About Tap To Pay` have different casing for "To" - maybe we can update `About Tap To Pay` to `About Tap to Pay` to be consistent.

**Comment by staskus** [comment]
Thanks for the review, @jaclync, I'll address the comments shortly.

> After tapping ‘Skip’, there’s no CTA Set up Tap to Pay, just Done. In another store, I was able to see Set Up Tap to Pay on iPhone CTA and the tap_to_pay_set_up_information_set_up_tapped event.

It depends if the store on device has TTP set up or not. If it doesn't, 'Set up Tap to Pay on iPhone' is shown. Otherwise, 'Done' is shown.

> I still couldn’t get the ToS to be shown again with a fresh install, https://businessconnect.apple.com/taptopay/removeall showed There are no merchant IDs associated with this Apple Account for my A8C Apple developer account. Do you see a different page?

This link shows me a list of merchant IDs that I can remove. Once I remove them, I can set up TTP again. However, this has to be the same account as on iPhone.

> During testing, I noticed an issue:
> Tap Not now --> after the view is dismissed, the Setup complete modal appears again which feels unexpected

Good observation, I'll address it!

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift (hunk 8 above)
Comment (staskus) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift]: I need this event tracked in two cases:
- When cancel is tapped
- When the view is dismissed with a drag-down gesture

Unfortunately, I had to do a lot of SwiftUI lifecycle gymnastics just to support both of these cases.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift (hunk 7 above)
Comment (staskus) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift]: It's true. I wanted to reuse the same events we had previously and contain all the receipt-related events in one place. I could move this event to the coordinator for the event to reflect the action.

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/BuiltInCardReaderConnectionController.swift (hunk 3 above)
Comment (staskus) [reply WooCommerce/Classes/ViewRelated/CardPresentPayments/BuiltInCardReaderConnectionController.swift]: Theoretically, users could try going into the location settings during the connection process and deny it but I don't want to handle such cases. Therefore, yes, it makes sense to limit this to only `denied` and `nonDetermined` cases.

File: Experiments/Experiments/DefaultFeatureFlagService.swift (hunk 1 above)
Comment (staskus) [reply Experiments/Experiments/DefaultFeatureFlagService.swift]: True, we could just enable the feature flags. I thought to do it by EOW but it can be done now as well.

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/Tap to Pay Education/TapToPayEducationView.swift (hunk 4 above)
Comment (staskus) [reply WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/Tap to Pay Education/TapToPayEducationView.swift:70]: Yes. Instead of duplicating `TapToPaySettingsFlowPresentingView` configuration and presentation, I'm passing a binding to `TapToPayEducationView` which presents this same view on the Payments view.

File: WooCommerce/WooCommerceTests/Tools/LocationServiceTests.swift (hunk 5 above)
Comment (staskus) [reply WooCommerce/WooCommerceTests/Tools/LocationServiceTests.swift:62]: Right now`requestPermission` method does less than before. Its only logic is calling `requestWhenInUseAuthorization()` when the status is `.notDetermined`. There's nothing else to test in that method.

The results are returned via the `observe` method which is called from the delegate methods. The logic is tested in the following tests.

File: Experiments/Experiments/DefaultFeatureFlagService.swift (hunk 1 above)
Comment (staskus) [reply Experiments/Experiments/DefaultFeatureFlagService.swift]: Addressed in https://github.com/woocommerce/woocommerce-ios/pull/14704/commits/b4afc8fcf43d88573d9ba3d650c9c06760daa963

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/BuiltInCardReaderConnectionController.swift (hunk 3 above)
Comment (staskus) [reply WooCommerce/Classes/ViewRelated/CardPresentPayments/BuiltInCardReaderConnectionController.swift]: Addressed in https://github.com/woocommerce/woocommerce-ios/pull/14704/commits/ad81d3598ac7548bcbc8fdf0e8eec99c168134b6

File: WooCommerce/WooCommerceTests/ViewModels/Receipts/ReceiptEmailViewModelTests.swift (hunk 6 above)
Comment (staskus) [reply WooCommerce/WooCommerceTests/ViewModels/Receipts/ReceiptEmailViewModelTests.swift]: Addressed in https://github.com/woocommerce/woocommerce-ios/pull/14704/commits/0941e016566726cbb1edeb2d903fc0363c17b86e

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift (hunk 7 above)
Comment (staskus) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift]: Addressed in https://github.com/woocommerce/woocommerce-ios/pull/14704/commits/14f5afd99c98df6aa67eb9b17323eb6bb1ac6f04. Moved the tracking in the coordinator as done with other receipt views.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift [hunk 11]
```diff
@@ -1,25 +1,32 @@
...
+    @Published private(set) var state: PrimaryLoadingButtonStyle.State = .idle
 
-    private let order: Order
+    private var order: Order
     private let stores: StoresManager
-    var noticePresenter: NoticePresenter
-    var emailValidator: (String) -> Bool = EmailFormatValidator.validate
-    var onDismiss: (Order?) -> Void
+    private var emailValidator: (String) -> Bool
+    private var onResult: (ReceiptEmailResult) -> Void
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Receipts/ReceiptEmail/ReceiptEmailViewModel.swift]: nit:
```suggestion
    private let emailValidator: (String) -> Bool
    private let onResult: (ReceiptEmailResult) -> Void
```

File: WooCommerce/Classes/ViewRelated/Orders/Collect Payments/CardPresentPaymentReceiptEmailCoordinator.swift [hunk 12]
```diff
@@ -58,6 +59,55 @@ final class CardPresentPaymentReceiptEmailCoordinator: NSObject {
...
+                    countryCode: countryCode,
+                    cardReaderModel: cardReaderModel,
+                    source: .api)
+                )
+                onCompleted(order)
+            case .failure(let error):
+                analytics.track(event: .InPersonPayments.receiptEmailFailed(
+                    error: error,
+                    countryCode: self.countryCode,
+                    cardReaderModel: self.cardReaderModel,
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/Orders/Collect Payments/CardPresentPaymentReceiptEmailCoordinator.swift]: super nit:
```suggestion
                    countryCode: countryCode,
                    cardReaderModel: cardReaderModel,
```

**Review by jaclync — APPROVED** [review]
Thanks for the updates, I tested the flow & side issue again and LGTM! :shipit:

**Comment by jaclync** [comment]
> > I still couldn’t get the ToS to be shown again with a fresh install, https://businessconnect.apple.com/taptopay/removeall showed There are no merchant IDs associated with this Apple Account for my A8C Apple developer account. Do you see a different page?
> 
> This link shows me a list of merchant IDs that I can remove. Once I remove them, I can set up TTP again. However, this has to be the same account as on iPhone.

Forgot to mention, logging into the right Apple account is key. I got to see and remove all the merchant IDs once I logged in to the personal Apple account on the device. Then I saw the ToS again.
